### PR TITLE
[DPE-5827] Set all nodes to synchronous replicas

### PR DESCRIPTION
**Pull request by dragomirp**

### CODE REVIEW

File: src/cluster.py [hunk 1]
```diff
@@ -846,7 +846,7 @@ def update_synchronous_node_count(self, units: int | None = None) -> None:
             with attempt:
                 r = requests.patch(
                     f"{self._patroni_url}/config",
-                    json={"synchronous_node_count": units // 2},
+                    json={"synchronous_node_count": units - 1},
```
Comment (dragomirp) [inline src/cluster.py:849]: REST calls should also use the new value.

**Review by lucasgameiroborges — APPROVED** [review]
Code LGTM, but why are we setting all nodes to be sync replicas? From the mentioned ticket it should come from a config value?

**Review by marceloneppel — CHANGES_REQUESTED** [review]
Cool! Testing it, it's working fine in almost all scenarios.

The scenario where it's not working yet is the upgrade:
```sh
juju deploy postgresql --channel 14/stable -n 3

# Wait for the units to settle down, then:
curl UNIT-ZERO-IP:8008/cluster | jq # There is one leader, one sync-standby and one replica, as expected.

juju run postgresql/leader pre-upgrade-check

juju refresh --path ./*.charm postgresql

# Wait for the units to settle down, then:
curl UNIT-ZERO-IP:8008/cluster | jq # There is one leader, one sync-standby and one replica, but we should have one leader and two sync-standby only.

juju add-unit postgresql -n 1

# Wait for the units to settle down, then:
curl UNIT-ZERO-IP:8008/cluster | jq # There is one leader, and three sync-standby, as expected.
```

**Review by taurus-forever — APPROVED** [review]
LGTM, need some time to play it without rush.
Let's merge it after release the current PostgreSQL VM to stable.

**Comment by dragomirp** [comment]
> The scenario where it's not working yet is the upgrade:

https://github.com/canonical/postgresql-operator/pull/672/commits/5082317f2f2eeed73b124966fc346424bb60c475 should fix the upgrade. Not adding an assertion to the integration tests because functionality will drift when the changes land on edge and stable.

File: src/cluster.py [hunk 2]
```diff
@@ -629,7 +629,7 @@ def render_patroni_yml_file(
             stanza=stanza,
             restore_stanza=restore_stanza,
             version=self.get_postgresql_version().split(".")[0],
-            minority_count=self.planned_units // 2,
+            synchronous_node_count=self.planned_units - 1,
```
[the remark this answers is not in the document]
Comment (nobuto-m) [reply src/cluster.py:650]: why don't you leave this information as a code comment?

[the remark this answers is not in the document]
Comment (dragomirp) [reply src/cluster.py:651]: Added as a comment.

**Comment by taurus-forever** [comment]
@dragomirp please do not forget to add config option `synchronous_node_count` as agreed in spec (with detailed description, the same as in spec).

**Comment by taurus-forever** [comment]
@dragomirp are we planning to include this PR the config option https://docs.google.com/document/d/1-a79fBYZPW2TUsKCGNyKjVWgLQefb9T7AiyV4QqW-y4/edit?tab=t.0 or it will be the separate PR?

**Comment by dragomirp** [comment]
> @dragomirp are we planning to include this PR the config option https://docs.google.com/document/d/1-a79fBYZPW2TUsKCGNyKjVWgLQefb9T7AiyV4QqW-y4/edit?tab=t.0 or it will be the separate PR?

I'll create a separate PR to this branch so all is in one place and reviewed, like I did for k8s.

**Review by marceloneppel — APPROVED** [review]
